src/Microsoft.VisualStudio.Editors/Common/ClientShimUtils.vb

```diff
@@ -0,0 +1,146 @@
+﻿' Licensed to the .NET Foundation under one or more agreements. The .NET Foundation licenses this file to you under the MIT license. See the LICENSE.md file in the project root for more information.
+
+Imports System.IO
+Imports Microsoft.VisualStudio.OLE.Interop
+Imports Microsoft.VisualStudio.ProjectSystem.Query
+Imports Microsoft.VisualStudio.ProjectSystem.Query.ProjectModel
+Imports Microsoft.VisualStudio.Shell.Interop
+Imports Microsoft.VisualStudio.ProjectSystem.Query.ProjectModel.Implementation
+Imports Microsoft.VisualStudio.ProjectSystem.Query.ServiceHub
+Imports Microsoft.VisualStudio.Shell
+
+Namespace Microsoft.VisualStudio.Editors.Common
+
+    ''' <summary>
+    ''' Shim layer to handle nexus and non-nexus APIs based on project
+    ''' </summary>
+    NotInheritable Class ClientShimUtils
+
+        Public Shared Function VSHierarchyShim(itemId As UInteger, VsHierarchy As IVsHierarchy, ServiceProvider As System.IServiceProvider) As IVsHierarchy
+           
+            'TODO - do we need to check if on nexus client ? If yes, how to check ?
+            ' If isClient = true Then
+            '   Return vsHierarchy
+            ' End If
+
+            ' Is Misc File and is nexus client then create the cps object
+            If IsMiscProjectFile(itemId, VsHierarchy) Then
+                Dim path As String
+                VsHierarchy.GetCanonicalName(itemId, path)
+                Return new ClientVsHierarchy(path, ServiceProvider)
+            End If
+
+            Return VsHierarchy
+        End Function
+
+        Private Shared Function IsMiscProjectFile(itemId As UInteger, vsHierarchy As IVsHierarchy) As Boolean
+
+            Dim pguid As Guid
+            vsHierarchy.GetGuidProperty(itemId, __VSHPROPID.VSHPROPID_ProjectIDGuid, pguid)
+
+            If pguid = VSConstants.CLSID.MiscellaneousFilesProject_guid Then
+                return True
+            End If
+
+            Return false
+        End Function
+
+    End Class
+
+    Friend Class ClientVsHierarchy
+        Implements IVsHierarchy
+
+        Private _nexusClientHierarchy As IVsHierarchy
+
+        Public Sub New(resxFilename As String, serviceProvider As System.IServiceProvider)
+
+            ThreadHelper.JoinableTaskFactory.Run(
+                Async Function() As Task(Of Task)
+                    Dim systemQueryService As IProjectSystemQueryService = CType(serviceProvider.GetService(GetType(IProjectSystemQueryService)), IProjectSystemQueryService)
+                    Dim queryableSpace = Await systemQueryService.GetProjectModelQueryableSpaceAsync()
+                    ' TODO : Exception. this operation is not valid
+                    Dim projectQuery = queryableSpace.Projects.With(Function(project) project.SourceFiles.With(Function(f) f.FileName = resxFilename))
```

Comment:
I think you want `Where` not `With`.

Reply:
To expand on that: `With` is used to select the data that will be returned by the query. `Where` is used to filter the query. So I think you want something like this instead:

``` VB
queryableSpace.Projects.With(Function(project) project.Name).With(Function(project) project.SourceFiles.Where(Function(source) source.FileName = resxFilename), isRequiredNotEmpty := True)
```

The first `With` says that you want the project name included in the results.

The second `With` says you want the source files in the project, but only the ones that match `resxFilename` (that's what `Where` does), and you only want projects where the filtered set of source files is not empty (`isRequiredNotEmpty := True`).

At least I think that's what you want. I'm basing this on the `TestQueryProjectWithSourceItemMatchingCondition` unit test in the CPS repo.

Reply:
This query also returns returns the same exception. This operation is not supported.

Reply:
I just confirmed with Lifeng, this is an issue that is reproducible in VB.